### DISCUSSION TITLE
1 주차 사전 과제, 유원종

### DISCUSSION THREAD

**Discussion by m20498** [MrMirror21/2024-KHU-CSE-TrackStudy-Web#16] [General]
- 명명 규칙(Naming Convention)이라는 것이 있습니다. 어떤 규칙들이 있는지, 각 규칙들은 어떻게 코드를 작성하는지, 자신이 그동안 사용해왔던 명명 규칙은 무엇인지 적어주세요.

Camel Case: 첫 단어는 소문자, 이후 단어의 첫 글자는 대문자.
예: myVariableName, getUserData()

Pascal Case: 모든 단어의 첫 글자는 대문자.
예: MyVariableName, GetUserData()

Snake Case: 모든 단어를 소문자로 쓰고, 단어 사이에 밑줄(_)을 넣음.
예: my_variable_name, get_user_data()

Upper Case Snake Case: 모든 글자를 대문자로 쓰고, 단어 사이에 밑줄(_)을 넣음.
예: MAX_VALUE, DEFAULT_TIMEOUT

개인적으로 사용한 명명 규칙
Python: Snake Case (my_variable)
JavaScript: Camel Case (myVariable)
상수: Upper Case Snake Case (MAX_VALUE)

- 함수가 순수하다는 것은 무엇을 의미할까요?

순수 함수는 다음 두 가지 조건을 만족하는 함수입니다:

항상 같은 입력에 대해 같은 출력: 입력 값이 같으면 항상 같은 결과를 반환.
Side Effect 없음: 함수 외부의 상태를 변경하지 않음.

- 개발 시에 순수 함수(pure function)로만 모든 코드를 작성하는 것에 대한 생각을 적어주세요.

순수 함수는 예측 가능하고 테스트가 쉽지만, 모든 코드를 순수 함수로만 작성하는 것은 어렵습니다. 파일 읽기/쓰기, 사용자 입력 등은 Side Effect가 필요하기 때문입니다. 따라서 순수 함수와 비순수 함수를 적절히 조합하여 사용하는 것이 좋습니다.

- 자바스크립트에서 함수를 작성하는 문법은 다음과 같습니다. 이 외의 함수를 작성하는 문법 두 가지가 무엇이며, 어떤 때 사용하나요?

```
function myFunc(param) {
	statements
}
```

1. 함수 표현식(Function Expression)

```
const myFunc = function(param) {
	// statement
};
```

동적 함수 정의: 함수 선언 시점이 실행 흐름 중간에 필요할 때.
익명 함수 사용: 이름이 필요 없는 함수로 콜백이나 일회성 작업에 사용.
블록 스코프: 특정 블록 내에서만 유효한 함수가 필요할 때.
호이스팅 방지: 함수 정의 전에 호출되는 것을 방지하고 싶을 때.

2. 화살표 함수(Arrow Function)

```
const myFunc = (param) => {
    // statements
};

// 호출
myFunc(arg);
```

간결한 문법: 짧고 간단한 함수가 필요할 때.
this 바인딩 필요 없음: 함수 내부의 this가 외부 스코프의 this를 유지해야 할 때.
콜백 함수: 짧은 콜백 함수 정의 시 유용.

- 파이썬에서는 문자열을 포맷팅하기 위해 다음과 같은 f-string을 사용합니다. 자바스크립트에서 f-string과 같이 문자열을 다양한 자료형과 변수들을 혼합하여 사용하려면 어떤 방법을 사용하나요?

아래 파이썬 코드 속 같은 조건과 같은 결과를 자바스크립트로 출력하면 됩니다.

```
>>> x = 1
>>> y = 2
>>> f"{x} + {y}는 {x + y}입니다."
'1 + 2는 3입니다.'
```

```
let x = 1;
let y = 2;
console.log(`${x} + ${y}는 ${x + y}입니다.`);
```

Template literals (템플릿 리터럴): 백틱(backtick) ` 문자로 문자열을 감싸고, ${} 구문을 사용하여 변수나 표현식을 삽입합니다.
${x}: 변수 x의 값을 문자열로 삽입합니다.
${y}: 변수 y의 값을 문자열로 삽입합니다.
${x + y}: 변수 x와 y의 합을 계산하여 문자열로 삽입합니다.

- 다음과 같은 객체(object)에서 p 값을 출력하려면 o.p 와 같은 방식으로 출력해야 하는데, p 로 출력할 수 있는 방법은 없을까요?

```
var o = { p: 42, q: true };

console.log(o.p); // 42
console.log(p); // p is undefined
```

```
var o = { p: 42, q: true };

// 객체의 프로퍼티를 전역 변수로 할당
var p = o.p;

console.log(o.p); // 42
console.log(p); // 42
```

**Comment by MrMirror21** [comment]
답변해주신 사전 과제 잘 봤습니다.
전체적으로 굉장히 훌륭하게 답변 잘해주셨고, 사이드 이펙트 개념까지 알고 계신 점이 인상적이네요.
마지막 문제의 경우 원종님이 답변해주신 방법도 해결책이 될 수 있으나, 리액트에서 자주 사용될 es6에서 도입된 문법이 있습니다. 스터디 중에 언급할테니 잘 들어주시면 도움이 될 것 같습니다. 👍
